WebKitGTK: rework how the minibrowser is downloaded on the CI (TaskCluster) and add weekly runs for beta channel

clopez opened the pull request:
Until now, when running the WebKitGTK tests on the taskcluster CI a pre-running hook was triggerred that installed some custom Debian packages for the stable channel and unpacked a tarball for the nightly channel. The tarball from the nightly channel required to install a lot of extra (uneeeded) dependencies (it was not optimized) slowing down the CI tests.

We recently re-worked how the pre-built products are generated, now we provide built products for stable, beta and nightly using the same bundle type: a zip file that contains just webkitgtk with the minimum libraries needed for it and is built for the target operating system where it will run. Currently we support Ubuntu LTS and LTS-1. TC CI runs in Ubuntu LTS.

This patch add supports for the command `wpt --install-browser` for product `webkitgtk_minibrowser` and uses this command to install the browser on the CI. The built product downloaded not longer requires using a specific path and is unpacked in the local directory.

This also means end-users can also install webkitgtk via this way and use it. But currently it will only work for them if they run Ubuntu LTS. We may add support for more distros in the future.

This patch also adds weekly runs for the beta channel.

The built products are available at https://webkitgtk.org/built-products/x86_64/release and automatic bots generate them daily.
The bots generating the built products for the nightly channel are at https://build.webkit.org and the ones for the stable and beta channels at https://build-webkit.igalia.com

clopez commented:
ping reviewers? //cc @jgraham

jgraham commented on tools/wpt/browser.py:
```diff
@@ -1311,7 +1311,100 @@ def version(self, binary=None, webdriver_binary=None):
+
+    def _find_executable_in_channel_bundle(self, binary, venv_path=None, channel=None):
+        if venv_path:
+            venv_base_path = self._get_dest(venv_path, channel)
```
Does this all rely on `dest` passed to `install` being the same as `venv_path`?

clopez replied:
Yes. It ends searching for the executable inside the folder ```_venv2/browsers/$channel/webkitgtk_minibrowser/```

jgraham replied:
Right, it seems unfortunate if we are basically assuming that "unrelated" functions are called with the same argument. We should at the very least document that this is a requirement. Ideally we could structure things so that we don't have to rely on this assumption because it's easy to foresee it breaking if we ever allow a user-specified download location or start using e.g. a temp directory for downloads.

clopez replied:
Sorry, but I'm not understanding your concern here. As far as I can see the all the other browser drivers relies on this same assumption regarding the functions to install and to find the browser binaries.

For example, the function to install the firefox browser gets the main subdirectory name to install the browser binaries via the [dest parameter passed to install ](https://github.com/web-platform-tests/wpt/blob/0305367bd0926faf7a2012d4cb7f82d092d8ec2c/tools/wpt/browser.py#L215) and the function to [find the browser binary first tries to find it on that directory via the parameter venv_path](https://github.com/web-platform-tests/wpt/blob/0305367bd0926faf7a2012d4cb7f82d092d8ec2c/tools/wpt/browser.py#L263).

As far as I can see the code ensuring the parameter passed is the same is this one: [here calls install](https://github.com/web-platform-tests/wpt/blob/0305367bd0926faf7a2012d4cb7f82d092d8ec2c/tools/wpt/run.py#L168) and [here calls find_binary](https://github.com/web-platform-tests/wpt/blob/0305367bd0926faf7a2012d4cb7f82d092d8ec2c/tools/wpt/run.py#L202).

I admit this is a bit confusing and the code overall for passing the parameters to install and find_binary may benefit from a refactoring, but I think that is out of the scope of this PR.

jgraham replied:
OK, fair enough. We should revisit this but in another PR.

clopez commented on tools/wpt/utils.py:
```diff
@@ -114,3 +122,46 @@ def handle_remove_readonly(func, path, exc):
     return shutil.rmtree(path, onerror=handle_remove_readonly)
+
+
+def download_url_to_descriptor(fd, url, max_retries=5):
```
I think both have different use cases: `download_url_to_descriptor` is meant to download big files (hundreds or even gigabytes of data) directly to the disk without using the memory as a cache. On the other hand `get()` is used as a simple way of obtaining the data into memory. It is used usually for inspecting json or small text files. Maybe renaming the function to something like `download_big_file_to_descriptor` would make it clear?

jgraham replied:
`get_download` or something perhaps, plus some docstrings to emphasise the difference.

clopez replied:
Ok. I will rename it and add some docstrings

clopez commented on tools/wpt/utils.py:
```diff
@@ -5,8 +5,16 @@
 import zipfile
 from io import BytesIO
+from socket import error as SocketError  # NOQA: N812
+try:
```
Good point!. Will change that.

clopez commented on tools/wpt/browser.py:
```diff
@@ -1323,15 +1416,13 @@ def find_binary(self, venv_path=None, channel=None):
-        if channel == "nightly":
-            path = "%s:%s" % (path, "/opt/webkitgtk/nightly")
-        return find_executable("WebKitWebDriver", path)
+    def find_webdriver(self, venv_path=None, channel=None):
```
Right... let's do it :)

clopez commented:
Pushed a new commit changing:
 - use `six.moves` in `tools/wpt/utils.py`
 - Properly update the function signature for `find_webdriver` and all the related classes

Please let me know if more changes are needed. Thanks!

clopez commented:
BTW.. I see 3 checks on the CI failed after the last fix-up commit (just a change on the pattern of the sha256checksum files) to accommodate to a [recent upstream webkit change](https://trac.webkit.org/changeset/266540). I don't think this 3 failures are valid.. after checking the log that seems to me like some random failure on that CI machines.

clopez committed:
WebKitGTK: rework how the minibrowser is downloaded on the CI (#25263)

 * Until now, when running the WebKitGTK tests on the taskcluster CI
a pre-running hook was triggerred that installed some custom Debian packages
for the stable channel and unpacked a tarball for the nightly channel.
The tarball from the nightly channel required to install a lot of extra
(uneeeded) dependencies (it was not optimized) slowing down the CI tests.

* We recently re-worked how the pre-built products are generated, now we
provide built products for stable, beta and nightly using the same bundle
type: a zip file that contains just webkitgtk with the minimum libraries
needed for it and is built for the target operating system where it will
run. Currently we support Ubuntu LTS and LTS-1. TC CI runs in Ubuntu LTS.

* This patch add supports for the command "wpt --install-browser" for product
webkitgtk_minibrowser and uses this command to install the browser on the CI.
The built product downloaded not longer requires using a specific path and
is unpacked in the local directory.

* This also means end-users can also install webkitgtk via this way and use it.
But currently it will only work for them if they run Ubuntu LTS.
We may add support for more distros in the future.

* Finally this commit also schedules weekly webkitgtk_minibrowser runs for the
beta channel on the CI.

clopez commented:
Sorry for the late replies, I was on holidays this past weeks.
I have checked the PR not longer applies on top of master (there were some conflicts due to 28bab3e887e which added a new function variant of ```find_webdriver()``` but that was easy to resolve).
Due to this i'm merging all the previous changes into one and force-pushing the rebased commit.
Sorry if this makes it a bit more difficult to continue the review. Regarding the previous version the main change is just:
 * rename download_url_to_descriptor() to get_download_to_descriptor(), improve the function docstrings to make clearer its use-case and update all the callers of it to use the new name.

clopez commented:
The failure on the CI seems unrelated to this commit. I tracked it down to be caused by https://github.com/web-platform-tests/wpt/pull/25774#issuecomment-700600138

clopez commented:
Rebased again on top of master to avoid the unrelated failure on the CI

clopez commented:
Everything green on the CI now

clopez commented:
Thanks!
